blockchain/utxocache.go

```diff
@@ -7,175 +7,64 @@ package blockchain
-//
-// This function is safe for concurrent access.
-func (ms *mapSlice) length() int {
-	ms.mtx.Lock()
-	defer ms.mtx.Unlock()
+	csm "github.com/mhmtszr/concurrent-swiss-map"
+)
 
-	var l int
-	for _, m := range ms.maps {
-		l += len(m)
+func NewCsMap(numMaxElements int) *csMap {
+	return &csMap{
+		maps: csm.Create(csm.WithSize[wire.OutPoint, *UtxoEntry](uint64(numMaxElements))),
 	}
-
-	return l
 }
 
-// size returns the size of all the maps in the map slice added together.
-//
-// This function is safe for concurrent access.
-func (ms *mapSlice) size() int {
-	ms.mtx.Lock()
-	defer ms.mtx.Unlock()
-
-	var size int
-	for _, num := range ms.maxEntries {
-		size += calculateRoughMapSize(num, bucketSize)
-	}
-
-	return size
+// csmap is a concurrent-swiss-map for UTXO entries.
+// csmap provides better control over memory allocation while ensuring efficient concurrent access.
+type csMap struct {
+	maps *csm.CsMap[wire.OutPoint, *UtxoEntry]
 }
 
-// get looks for the outpoint in all the maps in the map slice and returns
-// the entry.  nil and false is returned if the outpoint is not found.
+// length returns the length of csmap.
 //
 // This function is safe for concurrent access.
-func (ms *mapSlice) get(op wire.OutPoint) (*UtxoEntry, bool) {
-	ms.mtx.Lock()
-	defer ms.mtx.Unlock()
-
-	var entry *UtxoEntry
-	var found bool
-
-	for _, m := range ms.maps {
-		entry, found = m[op]
-		if found {
-			return entry, found
-		}
-	}
-
-	return nil, false
+func (ms *csMap) length() int {
+	return ms.maps.Count()
 }
 
-// put puts the outpoint and the entry into one of the maps in the map slice.  If the
-// existing maps are all full, it will allocate a new map based on how much memory we
-// have left over.  Leftover memory is calculated as:
-// maxTotalMemoryUsage - (totalEntryMemory + mapSlice.size())
+// size returns the size of csmap.
 //
 // This function is safe for concurrent access.
-func (ms *mapSlice) put(op wire.OutPoint, entry *UtxoEntry, totalEntryMemory uint64) {
-	ms.mtx.Lock()
-	defer ms.mtx.Unlock()
-
-	// Look for the key in the maps.
-	for i := range ms.maxEntries {
-		m := ms.maps[i]
-		_, found := m[op]
-		if found {
-			// If the key is found, overwrite it.
-			m[op] = entry
-			return // Return as we were successful in adding the entry.
-		}
-	}
-
-	for i, maxNum := range ms.maxEntries {
-		m := ms.maps[i]
-		if len(m) >= maxNum {
-			// Don't try to insert if the map already at max since
-			// that'll force the map to allocate double the memory it's
-			// currently taking up.
-			continue
-		}
-
-		m[op] = entry
-		return // Return as we were successful in adding the entry.
-	}
-
-	// We only reach this code if we've failed to insert into the map above as
-	// all the current maps were full.  We thus make a new map and insert into
-	// it.
-	m := ms.makeNewMap(totalEntryMemory)
-	m[op] = entry
+func (ms *csMap) size() int {
+	return calculateRoughMapSize(ms.length(), bucketSize)
```

Comment:
I think we also need to modify `calculateRoughMapSize`, as it makes assumptions w.r.t the current runtime implementation. With the introduction of Swiss maps in Go 1.24, this is likely out of date anyway (results in incorrect estimates).

Reply:
This right? 

https://go.dev/blog/swisstable

Maybe the map calculation might be easier now based on how the blog described it. We'll see...